pkg/xds/server/callbacks/dataplane_lifecycle.go

```diff
@@ -127,14 +126,16 @@ func (d *DataplaneLifecycle) register(
 			return errors.Wrap(err, "you are trying to override existing proxy to which you don't have an access.")
 		}
 		return existing.SetSpec(md.Resource.GetSpec())
-	},
-	)
+	})
 	if err != nil {
 		log.Info("cannot register proxy", "reason", err.Error())
+		if !loaded {
+			d.proxyInfos.Delete(proxyKey)
```

Comment:
I believe you need to indicate when the upsert succeeded in your proxyInfos as multiple instances could be stuck on acquiring the lock.

Reply:
Sorry, not sure I'm following this case. This method releases the lock in the end regardless of the result of the upsert. I don't see how other instances could be stuck

Reply:
It's the same as deregister where calls could be waiting for the lock (It might be fine though :) )